privacyidea/lib/crypto.py

```diff
@@ -811,16 +811,27 @@ def generate_otpkey(key_size=20):
 
 
 def generate_password(size=6, characters=string.ascii_lowercase +
-                        string.ascii_uppercase + string.digits):
+                        string.ascii_uppercase + string.digits, exclude='', requirements=[]):
     """
     Generate a random password of the specified length of the given characters
+    with optional requirements
 
     :param size: The length of the password
     :param characters: The characters the password may consist of
+    :param requirements: A list of strings from which the password must contain at least one character
+    :param exclude: A string of characters to exclude
     :return: password
     :rtype: basestring
     """
-    return ''.join(urandom.choice(characters) for _x in range(size))
+    # add one character from each string in the requirements list
+    passwd = [urandom.choice(str) for str in requirements]
```

Comment:
Please add a few tests for the method ``generate_password`` in test_lib_crypto.py, since this is rather unintuitive to me.

Reply:
added some tests